synapse/handlers/message.py

```diff
@@ -685,23 +704,40 @@ async def create_event(
 
         builder.internal_metadata.historical = historical
 
-        event, context = await self.create_new_client_event(
-            builder=builder,
-            requester=requester,
-            allow_no_prev_events=allow_no_prev_events,
-            prev_event_ids=prev_event_ids,
-            auth_event_ids=auth_event_ids,
-            state_event_ids=state_event_ids,
-            depth=depth,
-        )
+        if for_batch:
+            assert prev_event_ids is not None
+            assert state_map is not None
+            assert current_state_group is not None
+            auth_ids = self._event_auth_handler.compute_auth_events(builder, state_map)
+            event = await builder.build(
+                prev_event_ids=prev_event_ids, auth_event_ids=auth_ids, depth=depth
+            )
+            context = await self.state.compute_event_context_for_batched(
+                event, state_map, current_state_group
+            )
+
+        else:
+            event, context = await self.create_new_client_event(
+                builder=builder,
+                requester=requester,
+                allow_no_prev_events=allow_no_prev_events,
+                prev_event_ids=prev_event_ids,
+                auth_event_ids=auth_event_ids,
+                state_event_ids=state_event_ids,
+                depth=depth,
+            )
 
         # In an ideal world we wouldn't need the second part of this condition. However,
         # this behaviour isn't spec'd yet, meaning we should be able to deactivate this
         # behaviour. Another reason is that this code is also evaluated each time a new
         # m.room.aliases event is created, which includes hitting a /directory route.
         # Therefore not including this condition here would render the similar one in
         # synapse.handlers.directory pointless.
-        if builder.type == EventTypes.Aliases and self.require_membership_for_aliases:
+        if (
+            builder.type == EventTypes.Aliases
+            and self.require_membership_for_aliases
+            and not for_batch
```

Comment:
Why do we not do this for `for_batch`?

Reply:
My (admittedly poor) thinking on this was that since `for_batch` is only used with the createRoom codepath we already know that the sender is a member, but your question has made me realize this won't necessarily hold true if other code paths employ this mechanism. I have changed it so that the code now looks for the membership event id in the state map before trying to pull it from the DB if `for_batch` is true.